### DISCUSSION TITLE
Failed to install ruby 3.0.3 on Intel

### DISCUSSION THREAD

**Discussion by maximbond** [rbenv/ruby-build#1935] [Build failures]
Cannot install Ruby 3.0.3 on MacOS Monterey (same issue happened on Big Sur). rbenv 1.2.0, ruby-build 20220125.

### Output
```
$ RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.0.3
Downloading ruby-3.0.3.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.3.tar.gz
Installing ruby-3.0.3...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.2.1 using ruby-build 20220125)

Inspect or clean up the working tree at /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218165508.2693.xLyQt6
Results logged to /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218165508.2693.log

Last 10 log lines:
            ^
              _Nonnull
219 warnings and 2 errors generated.
make: *** [compile.o] Error 1
193 warnings generated.
219 warnings generated.
219 warnings generated.
194 warnings generated.
193 warnings generated.
193 warnings generated.
```
<img width="962" alt="Screenshot 2022-02-18 at 16 59 15" src="https://user-images.githubusercontent.com/7933971/154696311-9cd1d02e-7129-48a5-b111-debd6c65b1d7.png">

### Log
https://gist.github.com/maximbond/359df1bce1b9aaceed43ec203c32fb1e

### Why I added flags to rbenv install command
I added these flags because otherwise it pre-installs openssl-1.1.1l and it changes openssl to
```
openssl version
LibreSSL 2.8.3
```
After that I had to do this in order to get OpenSSL version back:
```
$ export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"
$ source ~/.zshrc
$ openssl version
OpenSSL 1.1.1m  14 Dec 2021
```

### Output 2
This is the output when I try to install Ruby w/o flags:
```
$ rbenv install 3.0.3

Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...
Installed openssl-1.1.1l to /Users/mks/.rbenv/versions/3.0.3

Downloading ruby-3.0.3.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.3.tar.gz
Installing ruby-3.0.3...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.2.1 using ruby-build 20220125)

Inspect or clean up the working tree at /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218165959.11004.WwWWTN
Results logged to /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218165959.11004.log

Last 10 log lines:
extern char *suboptarg;         /* getsubopt(3) external variable */
            ^
              _Nonnull
219 warnings and 2 errors generated.
make: *** [compile.o] Error 1
219 warnings generated.
219 warnings generated.
194 warnings generated.
193 warnings generated.
193 warnings generated.
```
<img width="990" alt="Screenshot 2022-02-18 at 17 13 13" src="https://user-images.githubusercontent.com/7933971/154698439-4ce75af1-798e-4f27-af99-ac054879e892.png">

### Log 2
It's huge! Almost 2m lines, gist didn't let me do that. I've compressed it and shared here: https://drive.google.com/file/d/15dZf2IM2HpUluMwdxtawYBNJXHTgTZCN/view?usp=sharing

**Comment by mislav** [comment]
Both the terminal output and the log that you've shared look incomplete. The terminal output says "Last 10 log lines", but what follows is not 10 lines. The full log that you've shared does not contain anything that looks like a compile error and does not even include the lines mentioned in "Last 10 log lines".

Consider using gist.github.com to share full logs instead of doing it inline, since logs can be huge.

**Comment by maximbond** [comment]
Sure @mislav - updated description and added full logs (+ screenshots). The second log didn't fit the gist however (130MB file), so I compressed it and shared via link. Not sure if it's a proper way to share logs in such cases, though.

**Reply by mislav** [reply]
The log file that you've shared:

> ### Log
> https://gist.github.com/maximbond/359df1bce1b9aaceed43ec203c32fb1e

Contains no errors/failures that I can see; only warnings. Are you sure you shared the whole file?

The number of warnings you are getting is worrying. I'm not sure what they mean, though. Have you tried upgrading Xcode, or installing command-line tools? `xcode-select --install` is a good starting point.

Is there a reason your CC environment variable seems to be `CC=clang -fdeclspec` and not just `CC=clang` or not set at all?

**Reply by maximbond** [reply]
@mislav  Try "view the full file" link at the top, I saw some errors like 
```
use of undeclared identifier 'RUBY_FUNCTION_NAME_STRING'
```

> Have you tried upgrading Xcode

I didn't have Xcode installed, just did it (from the App Store). 

> installing command-line tools

Already had them (should be the latest version I think? - I've re-installed them several times recently)
```
$ xcode-select --version
xcode-select version 2392.
```
> CC=clang -fdeclspec and not just CC=clang or not set at all

You are right, it wasn't set at all. Just set it in `.zshrc`, also added `openssl@1.1` to PATH in `.zshrc` and `.profile`.
Now it seems like they're set:
```
$ echo $CC
clang
$ openssl version
OpenSSL 1.1.1m  14 Dec 2021
```

So after that I tried to do install once again, but no luck and seems like errors are the same..
```
$ RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.0.3
Downloading ruby-3.0.3.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.3.tar.gz
Installing ruby-3.0.3...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.2.1 using ruby-build 20220125)

Inspect or clean up the working tree at /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218195827.76418.Cqri9U
Results logged to /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220218195827.76418.log

Last 10 log lines:
extern char *suboptarg;         /* getsubopt(3) external variable */
            ^
              _Nonnull
219 warnings and 2 errors generated.
make: *** [compile.o] Error 1
219 warnings generated.
219 warnings generated.
194 warnings generated.
193 warnings generated.
193 warnings generated.
```

### Log
https://gist.github.com/maximbond/beb602f239c582e61ba1eaf1cf121f30

**Reply by mislav** [reply]
@maximbond Thanks, I can see the log now. Try checking here to see if you get any tips: https://github.com/rbenv/ruby-build/issues/1431#issuecomment-617878504

**Reply by maximbond** [reply]
Hello @mislav! Thank you for this link - installation went much further, looks like previous errors are gone. Still an error though - but now it doesn't look that scary.

I've run:
```
brew update
brew upgrade
brew install llvm
export PATH="/usr/local/opt/llvm/bin:$PATH"
export LDFLAGS="-L/usr/local/opt/llvm/lib"
export CPPFLAGS="-I/usr/local/opt/llvm/include"
```
BTW ruby-build upgraded from 20220125  to 20220218

```
rbenv install 3.0.3
Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...

BUILD FAILED (macOS 12.2.1 using ruby-build 20220125)

Inspect or clean up the working tree at /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220223195309.76012.P5C9CE
Results logged to /var/folders/7_/240gy0092fj2gcm9bvzyls480000gp/T/ruby-build.20220223195309.76012.log

Last 10 log lines:
install ./include/openssl/uierr.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/uierr.h
install ./include/openssl/whrlpool.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/whrlpool.h
install ./include/openssl/x509.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/x509.h
install ./include/openssl/x509_vfy.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/x509_vfy.h
install ./include/openssl/x509err.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/x509err.h
install ./include/openssl/x509v3.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/x509v3.h
install ./include/openssl/x509v3err.h -> /Users/mks/.rbenv/versions/3.0.3/openssl/include/openssl/x509v3err.h
install libcrypto.a -> /Users/mks/.rbenv/versions/3.0.3/openssl/lib/libcrypto.a
ranlib: error: Invalid option: '-c'
make: *** [install_dev] Error 1
```
https://gist.github.com/maximbond/46be1d048364b770c061d453efa49d78

I feel like it's _almost fixed_ and all those issues are related to linking issues on my machine.
Also found [this tip](https://gist.github.com/aklap/e885721ef15c8668ed0a1dd64d2ea1a7) but a bit afraid to run those commands not to make things worse.

**Reply by maximbond** [reply]
Finally did it. So after all commands from my previous comment, I tried initial install command with flags and it worked:

1) Step 1
```
brew update
brew upgrade
brew install llvm
export PATH="/usr/local/opt/llvm/bin:$PATH"
export LDFLAGS="-L/usr/local/opt/llvm/lib"
export CPPFLAGS="-I/usr/local/opt/llvm/include"
```

2) Step 2
```
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.0.3
Downloading ruby-3.0.3.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.3.tar.gz
Installing ruby-3.0.3...
ruby-build: using readline from homebrew
Installed ruby-3.0.3 to /Users/mks/.rbenv/versions/3.0.3

➜  backend git:(master) ruby -v
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-darwin21]
```

Thank you @mislav. Thanks a lot.

**Reply by mislav** [reply]
I'm happy you solved it! But it shouldn't be necessary to explicitly install llvm via Homebrew to compile Ruby versions. I wonder what's wrong with the compiler that ships with Xcode on your system.
